abstract/encoding.go

```diff
@@ -170,13 +193,31 @@ func (de *decoder) value(v reflect.Value, depth int) error {
 			}
 		}
 
-	case reflect.Slice, reflect.Array:
+	case reflect.Slice:
+		if v.IsNil() {
+			panic("slices must be initialized to correct length before decoding")
+		}
+		fallthrough
+	case reflect.Array:
 		l := v.Len()
 		for i := 0; i < l; i++ {
 			if err = de.value(v.Index(i), depth+1); err != nil {
 				return err
 			}
 		}
+
+	case reflect.Int:
```

Comment:
Some possibilities:
  * always treat int as int32, but fail if we cut off the value -> faster on 32-bit architectures, but eventually failing when trying to transmit too big values
  * transmit as int64, try to cast to an int and fail if it's not possible -> faster on 64-bits, and if everybody is 64-bit, you can put bigger values, but might break once the app is rolled-out
  * log.Fail -> enforces usage of int32 or int64, but is complicated with lots of other libraries that might use int

A link for why we should allow for ints:
  * http://stackoverflow.com/questions/16427416/what-are-the-advantages-of-the-general-types-int-uint-over-specific-types-i